File: testprojects/csharp/SonarwhaleTestApi/Controllers/AuthController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using SonarwhaleTestApi.Models;

namespace SonarwhaleTestApi.Controllers;

/// <summary>Authentication — no auth required on any of these endpoints.</summary>
[ApiController]
[Route("api/[controller]")]
[AllowAnonymous]
public class AuthController : ControllerBase
{
    private const string SecretKey = "sonarwhale-test-secret-key-32chars!";

    /// <summary>Login with username and password. Returns JWT access + refresh token.</summary>
    /// <response code="200">Login successful</response>
    /// <response code="401">Invalid credentials</response>
    [HttpPost("login")]
    [ProducesResponseType(typeof(TokenResponse), 200)]
    [ProducesResponseType(401)]
    public IActionResult Login([FromBody] LoginRequest request)
    {
        // Hardcoded test users
        var users = new Dictionary<string, (string password, string role)>
        {
            ["admin"]    = ("admin123",    "admin"),
            ["user"]     = ("user123",     "user"),
            ["readonly"] = ("readonly123", "readonly"),
        };

        if (!users.TryGetValue(request.Username, out var info) || info.password != request.Password)
            return Unauthorized(new { message = "Invalid credentials" });

        var token = GenerateToken(request.Username, info.role);
        return Ok(new TokenResponse(token, $"refresh_{request.Username}_{Guid.NewGuid()}", 3600));
    }

    /// <summary>Exchange a refresh token for a new access token.</summary>
    /// <response code="200">New tokens issued</response>
    /// <response code="401">Invalid or expired refresh token</response>
    [HttpPost("refresh")]
    [ProducesResponseType(typeof(TokenResponse), 200)]
    [ProducesResponseType(401)]
    public IActionResult Refresh([FromBody] RefreshRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.RefreshToken) || !request.RefreshToken.StartsWith("refresh_"))
            return Unauthorized(new { message = "Invalid refresh token" });

        var parts = request.RefreshToken.Split('_');
        var username = parts.Length > 1 ? parts[1] : "user";
        var token = GenerateToken(username, "user");
        return Ok(new TokenResponse(token, $"refresh_{username}_{Guid.NewGuid()}", 3600));
    }

    /// <summary>Invalidate the current session / logout.</summary>
    /// <response code="204">Logged out successfully</response>
    [HttpPost("logout")]
    [ProducesResponseType(204)]
    public IActionResult Logout()
    {
        // In a real app: blacklist the token / clear the refresh token from DB
        return NoContent();
    }

    private static string GenerateToken(string username, string role)
    {
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(SecretKey));
        var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
        var claims = new[]
        {
            new Claim(ClaimTypes.Name, username),
            new Claim(ClaimTypes.Role, role),
            new Claim(JwtRegisteredClaimNames.Sub, username),
            new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString()),
        };
        var jwt = new JwtSecurityToken(
            issuer: "sonarwhale-test",
            audience: "sonarwhale-test",
            claims: claims,
            expires: DateTime.UtcNow.AddHours(1),
            signingCredentials: creds);
        return new JwtSecurityTokenHandler().WriteToken(jwt);
    }
}

File: testprojects/java/sonarwhale-test-api/src/main/java/dev/sonarwhale/testapi/controller/AuthController.java
package dev.sonarwhale.testapi.controller;

import dev.sonarwhale.testapi.model.LoginRequest;
import dev.sonarwhale.testapi.model.RefreshRequest;
import dev.sonarwhale.testapi.model.TokenResponse;
import dev.sonarwhale.testapi.security.JwtUtil;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

/**
 * Auth controller — no authentication required.
 *
 * JavaScanner coverage:
 *   @PostMapping("/login")   → POST /api/auth/login
 *   @PostMapping("/refresh") → POST /api/auth/refresh
 *   @PostMapping("/logout")  → POST /api/auth/logout
 */
@RestController
@RequestMapping("/api/auth")
@Tag(name = "Auth", description = "Authentication — no auth required")
public class AuthController {

    private static final Map<String, String[]> USERS = Map.of(
            "admin",    new String[]{"admin123",    "admin"},
            "user",     new String[]{"user123",     "user"},
            "readonly", new String[]{"readonly123", "readonly"}
    );

    private final JwtUtil jwtUtil;

    public AuthController(JwtUtil jwtUtil) {
        this.jwtUtil = jwtUtil;
    }

    @Operation(summary = "Login with username + password. Returns JWT access and refresh token.")
    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody LoginRequest request) {
        var user = USERS.get(request.username());
        if (user == null || !user[0].equals(request.password())) {
            return ResponseEntity.status(401).body(Map.of("message", "Invalid credentials"));
        }
        return ResponseEntity.ok(new TokenResponse(
                jwtUtil.generateToken(request.username(), user[1]),
                "refresh_" + request.username() + "_" + System.currentTimeMillis(),
                "bearer",
                3600));
    }

    @Operation(summary = "Exchange a refresh token for a new access token.")
    @PostMapping("/refresh")
    public ResponseEntity<?> refresh(@RequestBody RefreshRequest request) {
        if (!request.refreshToken().startsWith("refresh_")) {
            return ResponseEntity.status(401).body(Map.of("message", "Invalid refresh token"));
        }
        var parts    = request.refreshToken().split("_");
        var username = parts.length > 1 ? parts[1] : "user";
        var role     = USERS.containsKey(username) ? USERS.get(username)[1] : "user";
        return ResponseEntity.ok(new TokenResponse(
                jwtUtil.generateToken(username, role),
                "refresh_" + username + "_" + System.currentTimeMillis(),
                "bearer",
                3600));
    }

    @Operation(summary = "Invalidate the current session. Returns 204 No Content.")
    @PostMapping("/logout")
    public ResponseEntity<Void> logout() {
        // In a real app: add token to deny-list
        return ResponseEntity.noContent().build();
    }
}
